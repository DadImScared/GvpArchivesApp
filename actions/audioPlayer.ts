import { action } from "typesafe-actions";

import * as AudioActionTypes from "../actiontypes/audioPlayer";

export const togglePlaying = () => action(AudioActionTypes.TOGGLE_PLAYING);

export const setPlaying = (playing: boolean) => action(AudioActionTypes.SET_PLAYING, { playing });

export const showPlayer = (shouldShowPlayer: boolean) => {
  return action(AudioActionTypes.SHOW_PLAYER, { showPlayer: shouldShowPlayer });
};

export const setPlaylist = (name: string, items: string[], currentIndex = 0) => {
  return action(
    AudioActionTypes.SET_PLAYLIST,
    {
      currentIndex,
      items,
      name,
      playerType: "playlist" as any,
      playing: true,
      showPlayer: true,
    }
  );
};

export const updateIndex = (newIndex: number) => action(AudioActionTypes.UPDATE_CURRENT_INDEX, { newIndex });

export const setSong = (songUrl: string, songName: string) => {
  return action(
    AudioActionTypes.SET_SONG,
    {
      songName,
      songUrl,
      playerType: "song",
      playing: true,
      showPlayer: true
    }
  );
};

export interface IAudioPlayerActions {
  updateIndex: typeof updateIndex;
  togglePlaying: typeof togglePlaying;
  setSong: typeof setSong;
  setPlaying: typeof setPlaying;
  setPlaylist: typeof setPlaylist;
  showPlayer: typeof showPlayer;
}

export default {
  setPlaying,
  setPlaylist,
  setSong,
  showPlayer,
  togglePlaying,
  updateIndex,
};
