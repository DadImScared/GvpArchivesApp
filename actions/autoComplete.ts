import { action } from "typesafe-actions";

import * as ActionTypes from "../actiontypes/autoComplete";

const addAutoComplete = (query: string, categories: string[], results: string[]) => {
  return action(ActionTypes.ADD_AUTOCOMPLETE, { query, categories, results });
};

export default {
  addAutoComplete
};
