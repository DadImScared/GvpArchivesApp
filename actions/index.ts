export { default as autoComplete } from "./autoComplete";
export { default as audioPlayer } from "./audioPlayer";
export { default as itemsByCategory } from "./itemsByCategory";
export { default as itemsById } from "./itemsById";
export { default as loading } from "./loading";
export { default as search } from "./search";
