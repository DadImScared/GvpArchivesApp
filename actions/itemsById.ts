import { action } from "typesafe-actions";
import { ADD_ITEMS } from "../actiontypes/itemsById";
import { Item } from "../reducers/itemsById";

export const BASE_URL = __DEV__ ? "http://192.168.0.4:3000/api/v1" : "https://api.gvparchives.com/api/v1";

export const addItemsById = (items: { [key: string]: Item }) => action(ADD_ITEMS, { items });

export interface ItemsByIdActions {
  addItemsById: typeof addItemsById;
}

export default {
  addItemsById
};
