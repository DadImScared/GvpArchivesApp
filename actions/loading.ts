import { action } from "typesafe-actions";

import { LOADING_END, LOADING_START } from "../actiontypes/loading";

export const loadingStart = (id: string) => action(LOADING_START, id);

export const loadingEnd = (id: string) => action(LOADING_END, id);

export interface ILoadingActions {
  loadingStart: typeof loadingStart;
  loadingEnd: typeof loadingEnd;
}

export default {
  loadingEnd,
  loadingStart
};
