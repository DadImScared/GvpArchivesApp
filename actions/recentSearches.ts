import { action } from "typesafe-actions";

import { ADD_SEARCH } from "../actiontypes/recentSearches";

const addSearch = (query: string, categories: string[]) => {
  return action(ADD_SEARCH, { query, categories });
};

export default {
  addSearch
};
