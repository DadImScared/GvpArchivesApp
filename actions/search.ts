import axios, { AxiosPromise } from "axios";
import { AnyAction, Dispatch } from "redux";
import { action } from "typesafe-actions";

import * as ActionTypes from "../actiontypes/search";
import { AUTOCOMPLETE_QUERY } from "../loadingPrefix";
import { IReducerState } from "../reducers";
import { getQueryId } from "../reducers/search";
import autoComplete from "./autoComplete";
import { BASE_URL } from "./itemsById";

const autoCompleteUrl = `${BASE_URL}/api/v1/completeme/`;

export const isValidValidTimeToLive = (time: number) => time ? time >= +new Date() : false;

const updateSearchQuery = (query: string) => action(ActionTypes.UPDATE_SEARCH_QUERY, query);

type ApiCallCallback<ReturnType> = (
  state: IReducerState,
  dispatch: Dispatch
) => ReturnType;

type IAfterApiCall<Data> = (
  data: Data,
  state: IReducerState,
  dispatch: Dispatch
) => Promise<AnyAction[]> | AnyAction[];

export interface IApiCallConfig<Data> {
  afterApiCall?: IAfterApiCall<Data>;
  apiCall: ApiCallCallback<AxiosPromise<any>>;
  beforeApiCall?: ApiCallCallback<Promise<AnyAction[]> | AnyAction[]>;
  loadingId?: ((state: IReducerState) => string) | string;
  shouldCallApi?: ApiCallCallback<boolean | AxiosPromise<void>>;
}

export const getQuerySuggestions = (query: string, categories: string[]): IApiCallConfig<string[]> => {
  const queryId = getQueryId(query, categories);
  const loadingId = `${AUTOCOMPLETE_QUERY}${queryId}`;
  return {
    loadingId,
    afterApiCall: (data) => {
      return [
        autoComplete.addAutoComplete(
          query,
          categories,
          data.map((textObj: any) => textObj.text)
        )
      ];
    },
    apiCall: () => axios.get(`${autoCompleteUrl}${query}`),
    shouldCallApi: (state) => {
      const isLoading = state.loading[loadingId];
      if (!query || isLoading) return false;

      const currentAutoComplete = state.autoComplete[queryId];
      if (!currentAutoComplete) return true;

      return isValidValidTimeToLive(currentAutoComplete.timeToLive) ?
        false
        :
        axios.get(`${autoCompleteUrl}${query}`);
    }
  };
};

export default {
  updateSearchQuery
};
