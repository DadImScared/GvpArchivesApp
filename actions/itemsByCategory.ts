import axios from "axios";
import { normalize } from "normalizr";
import { Action, ActionCreator } from "redux";
import { batchActions } from "redux-batched-actions";
import { ThunkAction } from "redux-thunk";
import { action } from "typesafe-actions";
import { ADD_ITEMS_BY_CATEGORY, UPDATE_ITEMS_BY_CATEGORY } from "../actiontypes/itemsByCategory";
import { ITEMS_BY_CATEGORY } from "../loadingPrefix";
import { IReducerState } from "../reducers";
import { itemSchema } from "../schemas/items";
import { addItemsById, BASE_URL } from "./itemsById";
import { loadingEnd, loadingStart } from "./loading";

export type ActionCreatorThunk = ActionCreator<ThunkAction<void, IReducerState, undefined, Action<any>>>;

export const getItemsByCategory: ActionCreatorThunk = (category: string, nextPage?: string | boolean) => {
  return async (dispatch): Promise<void> => {
    const LOADING_ID = `${ITEMS_BY_CATEGORY}${category}`;
    dispatch(loadingStart(LOADING_ID));
    const firstPage = `${BASE_URL}/api/v1/items/${category}`;
    const url = typeof nextPage === "string" ? `${BASE_URL}${nextPage}` : firstPage;
    const response = await axios.get(url);
    if (response) {
      // dispatch batch actions
      const { data } = response;
      const { result: itemIds, entities: { items } } = normalize(data.results, itemSchema);
      dispatch(batchActions([
        addItemsById(items),
        (nextPage ? updateItemsByCategory : addItemsByCategory)({category, itemIds, nextPage: data.nextPage }),
        loadingEnd(LOADING_ID)
      ]));
    }
  };
};

interface ItemsByCategory {
  category: string;
  itemIds: string[];
  nextPage: boolean | string;
}

export const addItemsByCategory  = (data: ItemsByCategory) => {
  return action(ADD_ITEMS_BY_CATEGORY, data);
};

export const updateItemsByCategory = (data: ItemsByCategory) => {
  return action(UPDATE_ITEMS_BY_CATEGORY, data);
};

export interface ItemsByCategoryActions {
  addItemsByCategory: typeof addItemsByCategory;
  updateItemsByCategory: typeof updateItemsByCategory;
}

export default {
  addItemsByCategory,
  updateItemsByCategory
};
