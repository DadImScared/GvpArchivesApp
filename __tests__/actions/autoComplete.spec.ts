import { autoComplete } from "../../actions";

describe("autoComplete", () => {
  it("should add auto complete", () => {
    expect(autoComplete.addAutoComplete("query", ["book"], ["texthere"])).toMatchSnapshot();
  });
});
