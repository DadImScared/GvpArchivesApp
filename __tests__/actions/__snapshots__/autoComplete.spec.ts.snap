// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`autoComplete should add auto complete 1`] = `
Object {
  "meta": undefined,
  "payload": Object {
    "categories": Array [
      "book",
    ],
    "query": "query",
    "results": Array [
      "texthere",
    ],
  },
  "type": "autoComplete/ADD_AUTOCOMPLETE",
}
`;
